internal/dag/cache.go

```diff
@@ -146,11 +154,45 @@ func (kc *KubernetesCache) Insert(obj interface{}) bool {
 			kc.tlscertificatedelegations[k8s.NamespacedNameOf(obj)] = obj
 			return true
 		case *gatewayapi_v1alpha2.GatewayClass:
-			kc.gatewayclass = obj
-			return true
+			switch {
+			// Specific gateway configured: make sure the incoming gateway class
+			// matches that gateway's.
+			case kc.Gateway != nil:
+				if kc.gateway == nil || obj.Name != string(kc.gateway.Spec.GatewayClassName) {
+					return false
+				}
+
+				kc.gatewayclass = obj
+				return true
+			// Otherwise, take whatever we're given.
+			default:
+				kc.gatewayclass = obj
+				return true
+			}
 		case *gatewayapi_v1alpha2.Gateway:
-			kc.gateway = obj
-			return true
+			switch {
+			// Specific gateway configured: make sure the incoming gateway
+			// matches, and get its gateway class.
+			case kc.Gateway != nil:
+				if k8s.NamespacedNameOf(obj) != *kc.Gateway {
+					return false
+				}
+
+				kc.gateway = obj
+
+				gatewayClass := &gatewayapi_v1alpha2.GatewayClass{}
+				if err := kc.Client.Get(context.Background(), client.ObjectKey{Name: string(kc.gateway.Spec.GatewayClassName)}, gatewayClass); err != nil {
+					kc.WithError(err).Errorf("error getting gatewayclass for gateway %s/%s", kc.gateway.Namespace, kc.gateway.Name)
+				} else {
+					kc.gatewayclass = gatewayClass
+				}
```

Comment:
This pattern (fetch related resource B when resource A is added to the cache) is a little different than what we have elsewhere, but it seemed OK, particularly given the controller-runtime client will do reads from its cache, so we shouldn't actually be hitting the API server here, thoughts?

Reply:
This seems like a good approach to me, I think something like this pattern is common in other controller-runtime clients.